### PR TITLE
feat(acp): minimal generic ACP agent UI

### LINKED ISSUE #14222: feat(acp): minimal generic ACP agent UI (no per-provider scaffolding)
Replaces PR OpenHands/OpenHands#14189 with a minimal generic ACP UX. No per-provider (Claude Code / Codex / Gemini) scaffolding — the user supplies the subprocess command and env vars themselves. Feature-flagged on `enable_acp`.

## Settings page (`/settings/agent`)

Three controls, single tab:

1. **Agent type** — dropdown: OpenHands / ACP.
2. **Command** — textarea, whitespace-tokenized, required when ACP. Placeholder: `npx -y @agentclientprotocol/claude-agent-acp`.
3. **Environment variables** — `.env`-style textarea (one `KEY=value` per line, blanks and `#` ignored). Holds the API key and any base-URL override.
4. **Model** (optional) — text input → `acp_model`.

Save sends `{ agent_kind: \"acp\", acp_server: \"custom\", acp_command, acp_env, acp_model }`. Credentials stay in env (never on the command line).

## Sidebar gating when ACP is active

When `agent_settings.agent_kind === \"acp\"`, grey out (visible + disabled + tooltip):

| Item | Reason |
| -- | -- |
| LLM (`/settings`) | ACP subprocess manages its own LLM. |
| Condenser (`/settings/condenser`) | ACP server handles context. |
| MCP (`/settings/mcp`) | ACP server manages MCP. |

## Per-conversation ACP config snapshot

User's global ACP settings are defaults for **new** conversations only. Each conversation owns its own config — no mismatch is possible, so no mismatch banner.

* New encrypted `acp_settings: JSONB` column on `conversation_metadata` (alembic 010, idempotent).
* On start: snapshot the ACP fields into the column.
* On resume / webhook-import: hydrate `ACPAgent` from the snapshot, not from current user settings.
* Header shows `ACP: <last command token>` (e.g. `claude-agent-acp`), reusing the model-display element from OpenHands/OpenHands#13616 — no new component. Tooltip shows full command + args. Env never displayed.

## Feature flag and entry point

* `enable_acp: bool = False` in `WebClientFeatureFlags` and matching `ENABLE_ACP()` in `feature-flags.ts`.
* "Agent" item in the Settings sidebar, gated on `enable_acp`.

## Out of scope

* Per-provider convenience UX (named Claude Code / Codex / Gemini options).
* Auto-injecting SaaS proxy URL — users paste it. Add a pre-fill later if friction shows up.
* Editing a conversation's snapshotted config — frozen at start; rotate via a new conversation.

## Related

* Replaces PR OpenHands/OpenHands#14189; closes OpenHands/OpenHands#14183.
* Builds on backend in OpenHands/OpenHands#14004 and [#14171](<https://github.com/OpenHands/OpenHands/issues/14171>).

### PR DESCRIPTION
**Pull request by simonrosenberg**
- [x] A human has tested these changes.

---

## Why

ACP (Agent Client Protocol) agents run as external subprocesses rather than using OpenHands' built-in LLM stack. There was no UI to configure them — users had to edit \`settings.json\` by hand. This PR adds the minimal settings page needed to point OpenHands at an ACP agent, gates the rest of the settings UI so LLM/condenser/MCP options don't conflict while ACP is active, and wires up the routing + per-conversation provenance + UI plumbing.

It also bumps the SDK to **v1.22.1** which ships [software-agent-sdk#3227](https://github.com/OpenHands/software-agent-sdk/pull/3227) (unified ACP conversation endpoint), so OpenHands talks to a single \`/api/conversations\` route for both LLM and ACP agents instead of maintaining a parallel \`/api/acp/conversations\` dispatch.

Closes #14222 (the UI).

### Backend

- **SDK bump**: \`openhands-sdk\` / \`openhands-agent-server\` / \`openhands-tools\` pinned to **\`==1.22.1\`** (includes SDK PR #3227). The unified \`/api/conversations\` endpoint accepts both \`Agent\` and \`ACPAgent\` payloads through the \`AgentBase\` discriminated union, so OpenHands no longer dispatches by agent kind. \`AGENT_SERVER_IMAGE\` constant bumped to \`ghcr.io/openhands/agent-server:1.22.1-python\` to match.
- **Feature flag + registry**: \`enable_acp\` added to \`WebClientFeatureFlags\` (gated by \`ENABLE_ACP\`, default \`false\`). The SDK's \`ACP_PROVIDERS\` registry is mirrored as \`acp_providers\` on \`/api/v1/web-client/config\` so the frontend can resolve provider brand labels from one source.
- **No more route dispatch**: removed the \`agent_kind_to_router_path()\` helper, \`_split_ids_by_kind\` helper, the dual fetch in \`_get_live_conversation_info\`, and the per-call ACP-vs-LLM branches in \`live_status_app_conversation_service.py\`, \`remote_sandbox_service.py\`, and \`_start_conversation_remotely\`. Everything hits \`/api/conversations\`. \`StartACPConversationRequest\` / \`ACPConversationInfo\` references collapsed to the canonical \`StartConversationRequest\` / \`ConversationInfo\` (the ACP names are deprecated \`TypeAlias\` aliases on the new SDK).
- **Webhook router**: signature is \`ConversationInfo\` (the SDK's canonical model — \`ACPConversationInfo\` is now a deprecated alias). The \`AgentBase\` discriminated union accepts both \`Agent\` and \`ACPAgent\` payloads on the same route. \`on_conversation_update()\` sets \`agent_kind\` from the agent payload (\`llm_model\` for OpenHands, **null for ACP** — the ACP dummy LLM sentinel \`acp-managed\` is squashed to null so analytics dashboards don't get contaminated).
- **Per-conversation provider tag**: when an ACP conversation is created, the active provider key (\`"claude-code"\` / \`"codex"\` / \`"gemini-cli"\` / \`"custom"\`) is stamped onto \`AppConversationInfo.tags['acp_server']\` (inlined literal — no separate constant module). The frontend resolves the brand label ("Claude Code", "Codex", "Gemini CLI") from \`acp_providers\` at render time — single source of truth, no per-conversation column.
- **\`agent_kind\` column**: the column was added on \`main\` by #14004 (migration \`009\`, nullable \`String\`, no backfill needed). This PR doesn't change the schema — it just starts populating the column from webhook updates and start-task writes, and reads it on the frontend for the display label (with a \`'openhands'\` fallback for pre-#14004 rows).
- **Agent-kind switch (basic)**: switching \`agent_kind\` resets the new kind to defaults (avoids a mongrel deep-merge that would 500 the settings endpoint). Cross-kind config preservation (snapshot/restore) is **deferred** to a follow-up — switching back to OpenHands won't restore your previous LLM/condenser/MCP config.
- **\`agent_server_routing.py\` deleted**: the 5-line module existed only for the \`ACP_SERVER_TAG = 'acp_server'\` constant, which is now an inlined string literal — matching the frontend's already-inline treatment in \`agent-display-label.ts\`.
- **\`_load_skills_onto_request\` extracted**: the ACP and LLM arms of \`_build_start_conversation_request_for_user\` each had an identical try/except block calling \`_load_skills_and_update_agent\`. Pulled into one shared helper so skill-loading semantics only need to change in one place.
- **Lockfile bumps**: \`dev_config/python/.pre-commit-config.yaml\` (mypy SDK pin), root \`poetry.lock\`, \`uv.lock\`, and \`enterprise/poetry.lock\` regenerated to track the SDK v1.22.1 bump.

### Frontend

- **\`/settings/agent\` page** (\`routes/agent-settings.tsx\`): agent-type dropdown (OpenHands / ACP), preset selector — Claude Code / Codex / Gemini CLI / Custom — command textarea, optional model override. API keys go through the existing Secrets panel.
- **Route loader guard**: the \`clientLoader\` on LLM / Condenser / MCP settings pages redirects to \`/settings/agent\` when ACP is the active agent, preventing FOUC. No separate guard hook — the redirect is co-located with the route.
- **Nav gating**: LLM, Condenser, and MCP nav items are disabled (greyed out with tooltip) while ACP is active. The disabled-state prop carries the **agent name** (\`disabledAgentName\`), which the \`SETTINGS$AGENT_DISABLED_TOOLTIP\` i18n key interpolates as \`{agentName}\`.
- **Display label**: \`agentDisplayLabel()\` (in \`frontend/src/utils/agent-display-label.ts\`) resolves the short label shown next to a conversation title. For ACP it reads \`tags['acp_server']\`, looks the key up in \`acp_providers\`, and returns the brand name; falls back to \`"ACP"\` if the registry hasn't loaded or the key is unknown (custom commands). For OpenHands it returns \`"OpenHands"\` — the raw \`llm_model\` is preserved on hover via the \`title\` attribute.
- **Event rendering**: ACP tool-call cards are suppressed until \`status === "completed"\` or \`"failed"\` (prevents the empty-argument flash while \`raw_input\`/\`raw_output\` stream in). The card title is rendered verbatim from \`event.title\` — the upstream ACP sub-agent already emits a complete humanised label ("Read tests", "Edit foo.py"), so wrapping it in a translated verb prefix would double the verb.
- **WebSocket routing**: \`extractPathPrefix()\` matches the single \`/api/conversations\` shape since the SDK's \`RemoteConversation\` uses one route for both agent kinds now.
- **i18n**: 8 new \`SETTINGS$AGENT_*\` keys × 15 languages for the agent-settings page.

### Tests

- **\`tests/unit/app_server/test_webhook_router_acp.py\`** covers Agent vs ACPAgent discrimination on real Pydantic models (not MagicMock specs that would mask a discriminator regression), \`llm_model\` / \`agent_kind\` storage paths, \`tags['acp_server']\` preservation through a webhook update, and that \`track_conversation_created\` receives \`llm_model=None\` for ACP (regression guard against leaking the ACP \`LLM\` sentinel into analytics).
- **\`tests/unit/app_server/test_settings_agent_kind_switch.py\`** covers the OH ↔ ACP switch on \`Settings.update\`: it must not 500, the new kind comes up at defaults, \`acp_env\` is replaced wholesale, inline field overrides survive the switch, and \`mcp_config\` can be replaced in the same call.
- **\`TestAgentKindConversationUrl\`** in the live-status test file: parametrized regression guard for the unified \`/api/conversations/{id}\` URL — both \`openhands\` and \`acp\` agent kinds resolve to the same path. The \`_split_ids_by_kind\` and \`agent_kind_to_router_path\` test classes are deleted (the helpers no longer exist).
- **\`frontend/__tests__/routes/agent-settings.test.tsx\`** covers preset selection (incl. \`acp_server\` saved on submit) and switching back to OpenHands clearing ACP fields.
- Component tests verify the new chip resolution and the disabled-nav-item tooltip behaviour.

## How to Test

1. Enable ACP: set \`ENABLE_ACP=true\` on the backend.
2. Go to **Settings → Agent**, switch to **ACP**, pick the **Claude Code** preset (command auto-fills).
3. Add \`ANTHROPIC_API_KEY\` via **Settings → Secrets**.
4. Save. Start a new conversation — it should connect to the ACP subprocess and respond.
5. The conversation chip should read **"Claude Code"** (resolved from the \`acp_providers\` registry against the backend-stamped \`tags['acp_server']\`).
6. Verify **Settings → LLM**, **Condenser**, and **MCP** are greyed out in the nav while ACP is active, and redirect to \`/settings/agent\` if accessed directly.
7. Switch back to **OpenHands** — LLM/Condenser/MCP nav items become active again. (Note: your previous LLM/condenser/MCP config is **not** restored; the basic switch resets to defaults. Cross-kind preservation is a follow-up.)
8. Repeat the switch a few times to confirm no 500 errors.

Unit tests: \`cd frontend && npx vitest run\` and \`uv run pytest tests/unit/app_server/ -q\`.

## Type

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Breaking change
- [ ] Docs / chore

## Notes

- Feature is off by default. No existing behaviour changes until \`ENABLE_ACP=true\`.
- \`acp_env\` is supported at the API level but has no UI — credentials belong in the Secrets panel.
- SDK is pinned to **\`==1.22.1\`** (PyPI release). \`AGENT_SERVER_IMAGE\` points to the matching \`ghcr.io/openhands/agent-server:1.22.1-python\` build.
- The Settings → Agent preset dropdown is driven by \`acp_providers\` on \`/api/v1/web-client/config\` (the SDK registry), so adding a new provider on the backend shows up in the UI without code changes here.
- **Deferred follow-ups**:
  - **Cross-kind config preservation** (\`saved_agent_configs\` snapshot/restore on \`agent_kind\` switch). Today the new kind comes up at defaults.
  - **Custom-command resume after sandbox loss**: when a sandbox is destroyed (not just paused) and a user resumes an ACP conversation that used \`acp_server = custom\`, the \`acp_command\` / \`acp_args\` / \`acp_model\` come from the user's *current* settings, which may have changed. Built-in presets resume correctly because the SDK registry has \`default_command\`. A structured \`acp_spec\` field on \`AppConversationInfo\` would let custom-command conversations be reconstructed verbatim.

---

To run this PR locally:

```bash
ENABLE_ACP=true make run
```

Or natively (backend on :3000, frontend on :3001):

```bash
ENABLE_ACP=true make start-backend &
make start-frontend
```

The agent-server image (\`ghcr.io/openhands/agent-server:1.22.1-python\`) is already on GHCR and is what \`AGENT_SERVER_IMAGE\` resolves to by default.

---

To run this PR locally, use the following command:

GUI with Docker:
\`\`\`
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   --name openhands-app-4c08a87   docker.openhands.dev/openhands/openhands:4c08a87
\`\`\`

### CODE REVIEW

**Comment by simonrosenberg** [comment]
Status of the second-round review (10:23 UTC):

1. **Shell tokenization (blocker)** — Already addressed in 431d7b7 (pushed at 09:46 UTC, before this review round). `tokenizeCommand` is quote-aware (`npx -y "my pkg"` → `["npx", "-y", "my pkg"]`), `formatCommand` re-quotes tokens with whitespace so saved commands round-trip, and `frontend/__tests__/utils/shell-tokenize.test.ts` covers single quotes, double quotes, whitespace runs (incl. newlines), empty input, and a full round-trip. Backslash escapes / nested quotes are intentionally out of scope per the function's docstring — the textarea is not a shell.

2. **Hard-coded preset list** — False positive. The preset dropdown at `agent-settings.tsx:218-227` already consumes `acpProviders` from `/api/v1/web-client/config`:
   ```ts
   items={[
     ...acpProviders.map((p) => ({ key: p.key, label: p.display_name })),
     { key: CUSTOM_PRESET, label: t(I18nKey.SETTINGS$AGENT_PRESET_CUSTOM) },
   ]}
   ```
   The flag was driven by a stale `Notes` entry in the PR description that pre-dated the wiring; that entry has now been removed.

3. **Simplify initialization timing** — Marked "not blocking" in both reviews. The bot's Option A (no local state) doesn't fit a textarea that has to buffer keystrokes before save. Option B (`isSuccess` instead of `isLoading`/null-check) is mechanically equivalent here. The existing inline comments already document why `acpProviders` is in deps but doesn't re-trigger init.

**Comment by simonrosenberg** [comment]
Status of the third-round review (11:21 UTC):

1. **Shell tokenization "not visible in diff"** — Tooling artifact, not a code issue. The implementation file *is* part of the diff (`gh api repos/OpenHands/OpenHands/pulls/14401/files` confirms `frontend/src/utils/shell-tokenize.ts`, status=added, +55/-0). It was pushed in 431d7b7 alongside its tests. The functions are quote-aware and the test file exercises single quotes, double quotes, whitespace runs incl. newlines, empty input, and a round-trip.

2. **Migration not in diff** — Correct observation, misleading PR description. Migration `009` was actually shipped via #14004 (already on `main`); this PR only *populates and reads* the existing column. The column is nullable `String` with no backfill needed, reads use `stored.agent_kind or 'openhands'` as the fallback for pre-#14004 rows, and `agent_kind` isn't used in any `WHERE`/`ORDER BY`, so no index is warranted. PR description has been updated to make this clear.

3. **Agent switch loses user config** — Non-blocking and intentional. Documented in `How to Test` step 7 and in the `Deferred follow-ups` (cross-kind snapshot/restore). Adding inline messaging would mean a new i18n key × 15 languages for marginal value while the follow-up is still planned; skipping this round.

**Comment by simonrosenberg** [comment]
Re: [review 4281164712](https://github.com/OpenHands/OpenHands/pull/14401#pullrequestreview-4281164712) — pushed c87f2d4 to extend test coverage and tighten the init effect, and want to lay out the evidence on the three items so a human can verify rather than take my word for it.

**1. \"Tokenization implementation not visible in diff\"** — the implementation file *is* in the diff. The GitHub API returns its full 55-line patch directly:

\`\`\`
$ gh api repos/OpenHands/OpenHands/pulls/14401/files \\
    | jq '.[] | select(.filename | endswith(\"shell-tokenize.ts\"))
                | {filename, status, additions, patch_len: (.patch|length)}'
{
  \"filename\": \"frontend/src/utils/shell-tokenize.ts\",
  \"status\": \"added\",
  \"additions\": 55,
  \"patch_len\": 1642
}
\`\`\`

It was pushed in 431d7b7 (now superseded by c87f2d4). The bot view is probably truncating; the file is reachable in the [PR's Files Changed tab](https://github.com/OpenHands/OpenHands/pull/14401/files) and via the API.

c87f2d4 adds 4 more tests that lock down the edge cases the review enumerated:

- single vs. double quotes ✅ (existing)
- empty input / whitespace-only ✅ (existing)
- trailing whitespace ✅ (existing — handled by the \`inToken\` flag)
- backslash literal inside quotes ✅ (new — documents \"no escape semantics\")
- adjacent quoted segments concatenate into one token ✅ (new — matches POSIX shell behaviour)
- empty quoted segment yields \`\"\"\` token ✅ (new)
- unterminated quote is absorbed leniently into the trailing token ✅ (new)

12 tests total now, all passing. The docstring is explicit that backslash escapes and shell expansion are out of scope (this is a settings textarea, not a shell).

**2. \"Hard-coded preset list at agent-settings.tsx:158\"** — line 158 in the current file is the \`saveSettings(\` call, not a dropdown. The preset dropdown is at lines 218–227 and it already maps \`acpProviders\` exactly as the suggested code shows:

\`\`\`tsx
items={[
  ...acpProviders.map((provider) => ({
    key: provider.key,
    label: provider.display_name,
  })),
  { key: CUSTOM_PRESET, label: t(I18nKey.SETTINGS$AGENT_PRESET_CUSTOM) },
]}
\`\`\`

\`acpProviders = config?.acp_providers ?? EMPTY_ACP_PROVIDERS\` (line 55), so adding a provider in the SDK registry shows up in the UI with no frontend code change. The stale \"hardcoded follow-up\" entry in the PR description (which was the original trigger for this flag) has been removed.

**3. \"Initialization timing complexity\"** — addressed in c87f2d4. The two early returns are collapsed into one (\`if (!settings || isConfigLoading) return;\`) and the comment now spells out the timing contract: both queries must be settled, then the ref guard limits re-init to genuine \`settings\` identity changes (so an \`acpProviders\` refetch can't wipe in-progress edits).

**Comment by bastizol-wq** [comment]
On “feat(acp): minimal generic ACP agent UI”: you can get a lot of safety by doing a deterministic pre-flight before the agent calls the API.

If you already have OpenAPI/MCP metadata, you can make an allow/warn/block decision deterministically (and log it) before execution.

ATCP is an API-first gate that evaluates a tool/OpenAPI and returns a structured allow/warn/block result.

ATCP overview: https://www.sp-ai.io/atcp/

Curious what you’re using today for tool allowlists / safety receipts — happy to compare notes.

File: frontend/__tests__/utils/shell-tokenize.test.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    expect(tokenizeCommand('npx -y "my pkg"')).toEqual(["npx", "-y", "my pkg"]);
+  });
+
+  it("preserves arguments wrapped in single quotes", () => {
+    expect(tokenizeCommand("node 'path with spaces/cli.js' --acp")).toEqual([
+      "node",
+      "path with spaces/cli.js",
+      "--acp",
+    ]);
+  });
```
[the remark this answers is not in the document]
Comment (neubig) [reply frontend/__tests__/utils/shell-tokenize.test.ts:23]: This is indeed included now.

**Review by neubig — APPROVED** [review]
Looks good. Let's do this!
